storage: move ExternalIODir into StorageConfig

andrewbaptist opened the pull request:
Previously the ExternalIODir was stored as a variable inside the
BaseConfig struct and updated based on the StoreSpecs if it wasn't
explicitly set. This commit moves it into the StorageConfig and removes
it from the BaseConfig.

Epic: [CRDB-41111](https://cockroachlabs.atlassian.net/browse/CRDB-41111)

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/139980)
<!-- Reviewable:end -->

Commit:
storage: move ExternalIODir into StorageConfig

Previously the ExternalIODir was stored as a variable inside the
BaseConfig struct and updated based on the StoreSpecs if it wasn't
explicitly set. This commit moves it into the StorageConfig and removes
it from the BaseConfig.

Epic: CRDB-41111

Release note: None

andrewbaptist commented:
@RaduBerinde FYI - I'm investigating the failure in `TestChangefeedExecLocality` related to the change in `ServerArgsPerNode`.

andrewbaptist commented:
The failure in `TestChangefeedExecLocality` appears to be a random flake under stress+race and is unrelated to my change (other than by touching the file it caused CI to stress+race this file). Since it was a timeout rather than anything actionable, I'm going to ignore it for this PR's perspective.

RaduBerinde reviewed (commented):
I am not sure that ExternalIODir belongs in storage..  It is not directly related to the stores. @jbowens WDYT? This is the definition:
```
	// ExternalIODir is the local file path under which remotely-initiated
	// operations that can specify node-local I/O paths (such as BACKUP, RESTORE
	// or IMPORT) can access files.
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139980#-:-OHyHpTUAXrMGmzxIE5D:b2olgv9)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @golgeek, @herkolategan, @jbowens, @michae2, @msbutler, and @rharding6373)



<!-- Sent from Reviewable.io -->

jbowens reviewed (commented):
Yeah, agreed.. seems like it doesn't belong

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139980#-:-OIBOpyk3fxV-zz-fygP:b-20cvhw)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @golgeek, @herkolategan, @michae2, @msbutler, and @rharding6373)



<!-- Sent from Reviewable.io -->
